rpc_util_test.go

```diff
@@ -294,3 +298,89 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
 func BenchmarkGZIPCompressor1MiB(b *testing.B) {
 	bmCompressor(b, 1024*1024, NewGZIPCompressor())
 }
+
+// compressData compresses data using gzip and returns the compressed bytes.
+// It now accepts *testing.T to handle errors during compression.
+func compressData(t *testing.T, data []byte) []byte {
+	var buf bytes.Buffer
+	gz := gzip.NewWriter(&buf)
+	if _, err := gz.Write(data); err != nil {
+		t.Fatalf("compressData() failed to write data: %v", err)
+	}
+
+	if err := gz.Close(); err != nil {
+		t.Fatalf("compressData() failed to close gzip writer: %v", err)
+	}
+	return buf.Bytes()
+}
+
+// compressInput compresses input data and returns a BufferSlice.
+func compressInput(input []byte) mem.BufferSlice {
+	compressedData := compressData(nil, input)
+	return mem.BufferSlice{mem.NewBuffer(&compressedData, nil)}
+}
+
+// TestDecompress tests the decompress function with various scenarios, including
+// successful decompression, error handling, and edge cases like overflow or
+// premature data end. It ensures that the function behaves correctly with different
+// inputs, buffer sizes, and error conditions, using the "gzip" compressor for testing.
+func TestDecompress(t *testing.T) {
+	c := encoding.GetCompressor("gzip")
+
+	tests := []struct {
+		name                  string
+		compressor            encoding.Compressor
+		input                 []byte
+		maxReceiveMessageSize int
+		want                  []byte
+		wantErr               error
+	}{
+		{
+			name:                  "Decompresses successfully with sufficient buffer size",
+			compressor:            c,
+			input:                 []byte("decompressed data"),
+			maxReceiveMessageSize: 50,
+			want:                  []byte("decompressed data"),
+			wantErr:               nil,
+		},
+		{
+			name:                  "Fails due to exceeding maxReceiveMessageSize",
+			compressor:            c,
+			input:                 []byte("small message that is too large"),
+			maxReceiveMessageSize: 5,
+			want:                  nil,
+			wantErr:               errMaxMessageSizeExceeded,
+		},
+		{
+			name:                  "Decompresses to exactly maxReceiveMessageSize",
+			compressor:            c,
+			input:                 []byte("exact size message"),
+			maxReceiveMessageSize: len("exact size message"),
+			want:                  []byte("exact size message"),
+			wantErr:               nil,
+		},
+		{
+			name:                  "Handles maxReceiveMessageSize as MaxInt",
+			compressor:            c,
+			input:                 []byte("small message"),
+			maxReceiveMessageSize: math.MaxInt,
+			want:                  []byte("small message"),
+			wantErr:               nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			compressedMsg := compressInput(tt.input)
```

Comment:
We need to pass `t` as a param to `compressInput` so that it can call `t.Fatalf`. Right now, `compressInput` is passing a `nil` pointer to `compressData` which would cause a panic.